EventFilter/DTRawToDigi/plugins/DTuROSDigiToRaw.cc

```diff
@@ -0,0 +1,438 @@
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    lineFED+=4;
+    *((int*)lineFED)=dataWord2; 
+    lineFED+=4;
+    *((int*)lineFED)=dataWord1; 
+
+    dataWord1 = 0;
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    lineFED+=4;
+    *((int*)lineFED)=dataWord2; 
+    lineFED+=4;
+    *((int*)lineFED)=dataWord1; 
+
+    for (int nhit = 0; nhit < bslts[sltit]/2; nhit++) {
+
+      dataWord1 = 0x20000000
+                + wslts[sltit].at(nhit*2);
+      dataWord2 = wslts[sltit].at(nhit*2+1);
+
+      calcCRC(dataWord1, dataWord2, newCRC);
+      lineFED+=4;
+      *((int*)lineFED)=dataWord2; 
+      lineFED+=4;
+      *((int*)lineFED)=dataWord1; 
+    }
+
+    if (bslts[sltit]%2 == 1) {
+
+      dataWord1 = 0x20000000
+                + wslts[sltit].at(bslts[sltit]-1);
+      dataWord2 = 0x1FFFFFFF;
+
+      calcCRC(dataWord1, dataWord2, newCRC);
+      lineFED+=4;
+      *((int*)lineFED)=dataWord2; 
+      lineFED+=4;
+      *((int*)lineFED)=dataWord1; 
+    }
+
+    dataWord1 = 0x40000000;
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    lineFED+=4;
+    *((int*)lineFED)=dataWord2; 
+    lineFED+=4;
+    *((int*)lineFED)=dataWord1; 
+
+    dataWord1 = 0x40000000;
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    lineFED+=4;
+    *((int*)lineFED)=dataWord2; 
+    lineFED+=4;
+    *((int*)lineFED)=dataWord1; 
+
+    dataWord1 = 0;
+    dataWord2 = (dslts[sltit]&0xFFFFF);
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    lineFED+=4;
+    *((int*)lineFED)=dataWord2; 
+    lineFED+=4;
+    *((int*)lineFED)=dataWord1; 
+  }
+
+  //--> Trailer - line 1
+
+  dataWord1 = 0;
+  dataWord2 = 0;
+
+  calcCRC(dataWord1, dataWord2, newCRC);
+  lineFED+=4;
+  *((int*)lineFED)=dataWord2; 
+  lineFED+=4;
+  *((int*)lineFED)=dataWord1; 
+
+  //--> Trailer - line 2
+
+  dataWord1 = 0xA0000000
+             + (lines&0xFFFFFF);
+  dataWord2 = 0;
+
+  calcCRC(dataWord1, dataWord2&0xFFFF, newCRC);
+
+  dataWord2 += (newCRC&0xFFFF)<<16;
+
+  lineFED+=4;
+  *((int*)lineFED)=dataWord2; 
+  lineFED+=4;
+  *((int*)lineFED)=dataWord1; 
+
+  return;
+}
+
+
+void DTuROSDigiToRaw::clear() {
+
+  for (int sltit = 0; sltit < DOCESLOTS; sltit++) {
+
+    bslts[sltit]=0;
+    dslts[sltit]=0;
+    wslts[sltit].clear();
+  }
+
+  return;
+}
+
+
+int DTuROSDigiToRaw::theCRT(int ddu) {
+
+  if (ddu == 770) return 1368;
+  else if (ddu == 771) return 1368;
+  else if (ddu == 772) return 1369;
+  return 1370;
+}
+
+
+int DTuROSDigiToRaw::theSLT(int ddu, int ros, int rob) {
+
+  int slot = ((ros-1)/3)+1;
+  if (rob == 23) slot = 5;
+  if (ddu == 771) slot += 6;
+  else if (ddu == 773) slot += 6;
+  return slot;
+}
+
+
+int DTuROSDigiToRaw::theLNK(int ddu, int ros, int rob) {
+
+  int link = rob;
+  if (rob > 14) link = rob+1;
+  if (rob == 24) link = 15;
+  link += ((ros-1)%3)*24;
+  if (rob == 23) link = ros-1;
+  return link;
+}
+
+
+void DTuROSDigiToRaw::calcCRC(int myD1, int myD2, int & myC) {
```

Comment:
If I am not wrong, this function looks identical to 
https://cmssdt.cern.ch/lxr/source/EventFilter/DTTFRawToDigi/src/DTTFFEDSim.cc#0328 and
https://cmssdt.cern.ch/lxr/source/EventFilter/DTTFRawToDigi/src/DTTFFEDReader.cc#0524
If it has to be so by construction, you could avoid code duplication (and all risks of possible mistakes and bad maintenace connected to it) by defining one single helper function that will get called by all them.